### PR TITLE
Vrt valid hdr bench

### PR DESCRIPTION
**Pull request by nigoroll**
NOT FOR MERGE

benchmark for #3407 

* `configure.des`

```
$ for i in {1..10} ; do ./varnishtest -iv tests/bench_validhdr.vtc | grep VCL_Log ; echo; done
**** v1   vsl|       1001 VCL_Log         c bench /16 vct 0.249us
**** v1   vsl|       1001 VCL_Log         c bench /16 direct 0.296us
**** v1   vsl|       1002 VCL_Log         c bench /256 vct 1.009us
**** v1   vsl|       1002 VCL_Log         c bench /256 direct 0.858us
**** v1   vsl|       1003 VCL_Log         c bench /1024 vct 3.685us
**** v1   vsl|       1003 VCL_Log         c bench /1024 direct 2.794us

**** v1   vsl|       1001 VCL_Log         c bench /16 vct 0.218us
**** v1   vsl|       1001 VCL_Log         c bench /16 direct 0.196us
**** v1   vsl|       1002 VCL_Log         c bench /256 vct 1.175us
**** v1   vsl|       1002 VCL_Log         c bench /256 direct 0.854us
**** v1   vsl|       1003 VCL_Log         c bench /1024 vct 3.792us
**** v1   vsl|       1003 VCL_Log         c bench /1024 direct 2.726us

**** v1   vsl|       1001 VCL_Log         c bench /16 vct 0.220us
**** v1   vsl|       1001 VCL_Log         c bench /16 direct 0.194us
**** v1   vsl|       1002 VCL_Log         c bench /256 vct 1.087us
**** v1   vsl|       1002 VCL_Log         c bench /256 direct 0.907us
**** v1   vsl|       1003 VCL_Log         c bench /1024 vct 3.792us
**** v1   vsl|       1003 VCL_Log         c bench /1024 direct 2.748us

**** v1   vsl|       1001 VCL_Log         c bench /16 vct 0.214us
**** v1   vsl|       1001 VCL_Log         c bench /16 direct 0.193us
**** v1   vsl|       1002 VCL_Log         c bench /256 vct 1.008us
**** v1   vsl|       1002 VCL_Log         c bench /256 direct 0.834us
**** v1   vsl|       1003 VCL_Log         c bench /1024 vct 3.653us
**** v1   vsl|       1003 VCL_Log         c bench /1024 direct 2.878us

**** v1   vsl|       1001 VCL_Log         c bench /16 vct 0.215us
**** v1   vsl|       1001 VCL_Log         c bench /16 direct 0.194us
**** v1   vsl|       1002 VCL_Log         c bench /256 vct 1.007us
**** v1   vsl|       1002 VCL_Log         c bench /256 direct 0.821us
**** v1   vsl|       1003 VCL_Log         c bench /1024 vct 3.749us
**** v1   vsl|       1003 VCL_Log         c bench /1024 direct 2.785us

**** v1   vsl|       1001 VCL_Log         c bench /16 vct 0.215us
**** v1   vsl|       1001 VCL_Log         c bench /16 direct 0.192us
**** v1   vsl|       1002 VCL_Log         c bench /256 vct 1.006us
**** v1   vsl|       1002 VCL_Log         c bench /256 direct 0.861us
**** v1   vsl|       1003 VCL_Log         c bench /1024 vct 3.784us
**** v1   vsl|       1003 VCL_Log         c bench /1024 direct 2.685us

**** v1   vsl|       1001 VCL_Log         c bench /16 vct 0.215us
**** v1   vsl|       1001 VCL_Log         c bench /16 direct 0.193us
**** v1   vsl|       1002 VCL_Log         c bench /256 vct 1.075us
**** v1   vsl|       1002 VCL_Log         c bench /256 direct 0.949us
**** v1   vsl|       1003 VCL_Log         c bench /1024 vct 3.718us
**** v1   vsl|       1003 VCL_Log         c bench /1024 direct 2.657us

**** v1   vsl|       1001 VCL_Log         c bench /16 vct 0.215us
**** v1   vsl|       1001 VCL_Log         c bench /16 direct 0.193us
**** v1   vsl|       1002 VCL_Log         c bench /256 vct 1.006us
**** v1   vsl|       1002 VCL_Log         c bench /256 direct 0.840us
**** v1   vsl|       1003 VCL_Log         c bench /1024 vct 3.704us
**** v1   vsl|       1003 VCL_Log         c bench /1024 direct 2.738us

**** v1   vsl|       1001 VCL_Log         c bench /16 vct 0.214us
**** v1   vsl|       1001 VCL_Log         c bench /16 direct 0.192us
**** v1   vsl|       1002 VCL_Log         c bench /256 vct 1.007us
**** v1   vsl|       1002 VCL_Log         c bench /256 direct 0.956us
**** v1   vsl|       1003 VCL_Log         c bench /1024 vct 3.773us
**** v1   vsl|       1003 VCL_Log         c bench /1024 direct 2.606us

**** v1   vsl|       1001 VCL_Log         c bench /16 vct 0.217us
**** v1   vsl|       1001 VCL_Log         c bench /16 direct 0.196us
**** v1   vsl|       1002 VCL_Log         c bench /256 vct 1.010us
**** v1   vsl|       1002 VCL_Log         c bench /256 direct 0.840us
**** v1   vsl|       1003 VCL_Log         c bench /1024 vct 3.725us
**** v1   vsl|       1003 VCL_Log         c bench /1024 direct 2.667us
```

* `./configure CFLAGS="-O6"`

```
$ for i in {1..10} ; do ./varnishtest -iv tests/bench_validhdr.vtc | grep VCL_Log ; echo; done
**** v1   vsl|       1001 VCL_Log         c bench /16 vct 0.056us
**** v1   vsl|       1001 VCL_Log         c bench /16 direct 0.054us
**** v1   vsl|       1002 VCL_Log         c bench /256 vct 0.159us
**** v1   vsl|       1002 VCL_Log         c bench /256 direct 0.170us
**** v1   vsl|       1003 VCL_Log         c bench /1024 vct 0.443us
**** v1   vsl|       1003 VCL_Log         c bench /1024 direct 0.496us

**** v1   vsl|       1001 VCL_Log         c bench /16 vct 0.056us
**** v1   vsl|       1001 VCL_Log         c bench /16 direct 0.055us
**** v1   vsl|       1002 VCL_Log         c bench /256 vct 0.159us
**** v1   vsl|       1002 VCL_Log         c bench /256 direct 0.170us
**** v1   vsl|       1003 VCL_Log         c bench /1024 vct 0.442us
**** v1   vsl|       1003 VCL_Log         c bench /1024 direct 0.496us

**** v1   vsl|       1001 VCL_Log         c bench /16 vct 0.055us
**** v1   vsl|       1001 VCL_Log         c bench /16 direct 0.054us
**** v1   vsl|       1002 VCL_Log         c bench /256 vct 0.163us
**** v1   vsl|       1002 VCL_Log         c bench /256 direct 0.172us
**** v1   vsl|       1003 VCL_Log         c bench /1024 vct 0.469us
**** v1   vsl|       1003 VCL_Log         c bench /1024 direct 0.554us

**** v1   vsl|       1001 VCL_Log         c bench /16 vct 0.114us
**** v1   vsl|       1001 VCL_Log         c bench /16 direct 0.117us
**** v1   vsl|       1002 VCL_Log         c bench /256 vct 0.338us
**** v1   vsl|       1002 VCL_Log         c bench /256 direct 0.359us
**** v1   vsl|       1003 VCL_Log         c bench /1024 vct 0.981us
**** v1   vsl|       1003 VCL_Log         c bench /1024 direct 1.297us

**** v1   vsl|       1001 VCL_Log         c bench /16 vct 0.063us
**** v1   vsl|       1001 VCL_Log         c bench /16 direct 0.063us
**** v1   vsl|       1002 VCL_Log         c bench /256 vct 0.217us
**** v1   vsl|       1002 VCL_Log         c bench /256 direct 0.170us
**** v1   vsl|       1003 VCL_Log         c bench /1024 vct 0.443us
**** v1   vsl|       1003 VCL_Log         c bench /1024 direct 0.497us

**** v1   vsl|       1001 VCL_Log         c bench /16 vct 0.055us
**** v1   vsl|       1001 VCL_Log         c bench /16 direct 0.055us
**** v1   vsl|       1002 VCL_Log         c bench /256 vct 0.159us
**** v1   vsl|       1002 VCL_Log         c bench /256 direct 0.169us
**** v1   vsl|       1003 VCL_Log         c bench /1024 vct 0.442us
**** v1   vsl|       1003 VCL_Log         c bench /1024 direct 0.496us

**** v1   vsl|       1001 VCL_Log         c bench /16 vct 0.056us
**** v1   vsl|       1001 VCL_Log         c bench /16 direct 0.054us
**** v1   vsl|       1002 VCL_Log         c bench /256 vct 0.159us
**** v1   vsl|       1002 VCL_Log         c bench /256 direct 0.170us
**** v1   vsl|       1003 VCL_Log         c bench /1024 vct 0.443us
**** v1   vsl|       1003 VCL_Log         c bench /1024 direct 0.496us

**** v1   vsl|       1001 VCL_Log         c bench /16 vct 0.055us
**** v1   vsl|       1001 VCL_Log         c bench /16 direct 0.055us
**** v1   vsl|       1002 VCL_Log         c bench /256 vct 0.159us
**** v1   vsl|       1002 VCL_Log         c bench /256 direct 0.169us
**** v1   vsl|       1003 VCL_Log         c bench /1024 vct 0.442us
**** v1   vsl|       1003 VCL_Log         c bench /1024 direct 0.497us

**** v1   vsl|       1001 VCL_Log         c bench /16 vct 0.055us
**** v1   vsl|       1001 VCL_Log         c bench /16 direct 0.054us
**** v1   vsl|       1002 VCL_Log         c bench /256 vct 0.159us
**** v1   vsl|       1002 VCL_Log         c bench /256 direct 0.170us
**** v1   vsl|       1003 VCL_Log         c bench /1024 vct 0.459us
**** v1   vsl|       1003 VCL_Log         c bench /1024 direct 0.496us

**** v1   vsl|       1001 VCL_Log         c bench /16 vct 0.056us
**** v1   vsl|       1001 VCL_Log         c bench /16 direct 0.055us
**** v1   vsl|       1002 VCL_Log         c bench /256 vct 0.159us
**** v1   vsl|       1002 VCL_Log         c bench /256 direct 0.169us
**** v1   vsl|       1003 VCL_Log         c bench /1024 vct 0.443us
**** v1   vsl|       1003 VCL_Log         c bench /1024 direct 0.497us
```

### CODE REVIEW

File: bin/varnishd/cache/cache_vrt.c [hunk 1]
```diff
@@ -590,6 +590,61 @@ VRT_UpperLowerStrands(VRT_CTX, VCL_STRANDS s, int up)
...
 }
 
+// RFC7232, 3.2 without obsolete line folding:
+// ASCII VCHAR + TAB + obs-text (0x80-ff)
+static inline VCL_BOOL
+validhdr_vct(const char *p)
+{
+	AN(p);
+	for(;*p != '\0'; p++)
+		if (vct_isctl(*p) && !vct_issp(*p))
```
Comment (Dridi) [inline bin/varnishd/cache/cache_vrt.c:600]: Didn't we agree that we should have `VCT_OBSTEXT` and `vct_ishdrval()` for the bench?

File: bin/varnishd/cache/cache_vrt.c [hunk 2]
```diff
@@ -590,6 +590,61 @@ VRT_UpperLowerStrands(VRT_CTX, VCL_STRANDS s, int up)
...
+			return (0);
+	return (1);
+}
+#define vct_ishdrval(x) (((x) >= 0x20 && (x) != 0x7f) || (x) == 0x09)
```
Comment (Dridi) [inline bin/varnishd/cache/cache_vrt.c:604]: Is it safe not to cast `x` as `uint8_t` or similar?

**Review by Dridi — CHANGES_REQUESTED** [review]
Please rebase the PR on top of master instead of #3163.

**Comment by nigoroll** [comment]
> Please rebase the PR on top of master instead of #3163.
I would do that if master had the benchmark facility, which is based on `VRT_call()`

**Comment by Dridi** [comment]
> I would do that if master had the benchmark facility, which is based on `VRT_call()`

Never mind then :)

**Comment by nigoroll** [comment]
note: I am having a hard time grappling with the `-O6` results. When I benchmark the code with another variant added, the `vct` results look completely different, even though the code is bit-identical (as shown with `objdump -S`).

**Comment by nigoroll** [comment]
my results for `configure.des` with the latest push 3b336773af7e57d4ad977096ea0256ae52f9cd2c

```
$ for i in {1..10} ; do ./varnishtest -iv tests/bench_validhdr.vtc | grep VCL_Log ; echo; done
**** v1   vsl|       1001 VCL_Log         c bench /16 vct 0.220us
**** v1   vsl|       1001 VCL_Log         c bench /16 vcthdrval 0.220us
**** v1   vsl|       1001 VCL_Log         c bench /16 direct 0.197us
**** v1   vsl|       1002 VCL_Log         c bench /256 vct 0.941us
**** v1   vsl|       1002 VCL_Log         c bench /256 vcthdrval 1.088us
**** v1   vsl|       1002 VCL_Log         c bench /256 direct 0.843us
**** v1   vsl|       1003 VCL_Log         c bench /1024 vct 3.241us
**** v1   vsl|       1003 VCL_Log         c bench /1024 vcthdrval 3.728us
**** v1   vsl|       1003 VCL_Log         c bench /1024 direct 2.833us

**** v1   vsl|       1001 VCL_Log         c bench /16 vct 0.218us
**** v1   vsl|       1001 VCL_Log         c bench /16 vcthdrval 0.221us
**** v1   vsl|       1001 VCL_Log         c bench /16 direct 0.197us
**** v1   vsl|       1002 VCL_Log         c bench /256 vct 0.942us
**** v1   vsl|       1002 VCL_Log         c bench /256 vcthdrval 1.050us
**** v1   vsl|       1002 VCL_Log         c bench /256 direct 0.790us
**** v1   vsl|       1003 VCL_Log         c bench /1024 vct 3.300us
**** v1   vsl|       1003 VCL_Log         c bench /1024 vcthdrval 3.753us
**** v1   vsl|       1003 VCL_Log         c bench /1024 direct 2.804us

**** v1   vsl|       1001 VCL_Log         c bench /16 vct 0.235us
**** v1   vsl|       1001 VCL_Log         c bench /16 vcthdrval 0.240us
**** v1   vsl|       1001 VCL_Log         c bench /16 direct 0.216us
**** v1   vsl|       1002 VCL_Log         c bench /256 vct 0.964us
**** v1   vsl|       1002 VCL_Log         c bench /256 vcthdrval 1.061us
**** v1   vsl|       1002 VCL_Log         c bench /256 direct 0.844us
**** v1   vsl|       1003 VCL_Log         c bench /1024 vct 3.286us
**** v1   vsl|       1003 VCL_Log         c bench /1024 vcthdrval 4.042us
**** v1   vsl|       1003 VCL_Log         c bench /1024 direct 2.697us

**** v1   vsl|       1001 VCL_Log         c bench /16 vct 0.308us
**** v1   vsl|       1001 VCL_Log         c bench /16 vcthdrval 0.220us
**** v1   vsl|       1001 VCL_Log         c bench /16 direct 0.196us
**** v1   vsl|       1002 VCL_Log         c bench /256 vct 0.974us
**** v1   vsl|       1002 VCL_Log         c bench /256 vcthdrval 1.031us
**** v1   vsl|       1002 VCL_Log         c bench /256 direct 0.790us
**** v1   vsl|       1003 VCL_Log         c bench /1024 vct 3.268us
**** v1   vsl|       1003 VCL_Log         c bench /1024 vcthdrval 3.805us
**** v1   vsl|       1003 VCL_Log         c bench /1024 direct 2.755us

**** v1   vsl|       1001 VCL_Log         c bench /16 vct 0.216us
**** v1   vsl|       1001 VCL_Log         c bench /16 vcthdrval 0.221us
**** v1   vsl|       1001 VCL_Log         c bench /16 direct 0.197us
**** v1   vsl|       1002 VCL_Log         c bench /256 vct 0.943us
**** v1   vsl|       1002 VCL_Log         c bench /256 vcthdrval 1.080us
**** v1   vsl|       1002 VCL_Log         c bench /256 direct 0.842us
**** v1   vsl|       1003 VCL_Log         c bench /1024 vct 3.237us
**** v1   vsl|       1003 VCL_Log         c bench /1024 vcthdrval 3.648us
**** v1   vsl|       1003 VCL_Log         c bench /1024 direct 2.755us

**** v1   vsl|       1001 VCL_Log         c bench /16 vct 0.220us
**** v1   vsl|       1001 VCL_Log         c bench /16 vcthdrval 0.224us
**** v1   vsl|       1001 VCL_Log         c bench /16 direct 0.200us
**** v1   vsl|       1002 VCL_Log         c bench /256 vct 1.025us
**** v1   vsl|       1002 VCL_Log         c bench /256 vcthdrval 1.165us
**** v1   vsl|       1002 VCL_Log         c bench /256 direct 0.850us
**** v1   vsl|       1003 VCL_Log         c bench /1024 vct 3.355us
**** v1   vsl|       1003 VCL_Log         c bench /1024 vcthdrval 3.657us
**** v1   vsl|       1003 VCL_Log         c bench /1024 direct 2.721us

**** v1   vsl|       1001 VCL_Log         c bench /16 vct 0.438us
**** v1   vsl|       1001 VCL_Log         c bench /16 vcthdrval 0.482us
**** v1   vsl|       1001 VCL_Log         c bench /16 direct 0.381us
**** v1   vsl|       1002 VCL_Log         c bench /256 vct 2.040us
**** v1   vsl|       1002 VCL_Log         c bench /256 vcthdrval 2.448us
**** v1   vsl|       1002 VCL_Log         c bench /256 direct 1.141us
**** v1   vsl|       1003 VCL_Log         c bench /1024 vct 7.092us
**** v1   vsl|       1003 VCL_Log         c bench /1024 vcthdrval 8.655us
**** v1   vsl|       1003 VCL_Log         c bench /1024 direct 3.369us

**** v1   vsl|       1001 VCL_Log         c bench /16 vct 0.216us
**** v1   vsl|       1001 VCL_Log         c bench /16 vcthdrval 0.222us
**** v1   vsl|       1001 VCL_Log         c bench /16 direct 0.197us
**** v1   vsl|       1002 VCL_Log         c bench /256 vct 0.973us
**** v1   vsl|       1002 VCL_Log         c bench /256 vcthdrval 1.034us
**** v1   vsl|       1002 VCL_Log         c bench /256 direct 0.843us
**** v1   vsl|       1003 VCL_Log         c bench /1024 vct 3.281us
**** v1   vsl|       1003 VCL_Log         c bench /1024 vcthdrval 5.485us
**** v1   vsl|       1003 VCL_Log         c bench /1024 direct 2.854us

**** v1   vsl|       1001 VCL_Log         c bench /16 vct 0.217us
**** v1   vsl|       1001 VCL_Log         c bench /16 vcthdrval 0.221us
**** v1   vsl|       1001 VCL_Log         c bench /16 direct 0.197us
**** v1   vsl|       1002 VCL_Log         c bench /256 vct 0.946us
**** v1   vsl|       1002 VCL_Log         c bench /256 vcthdrval 1.075us
**** v1   vsl|       1002 VCL_Log         c bench /256 direct 0.842us
**** v1   vsl|       1003 VCL_Log         c bench /1024 vct 3.302us
**** v1   vsl|       1003 VCL_Log         c bench /1024 vcthdrval 3.686us
**** v1   vsl|       1003 VCL_Log         c bench /1024 direct 2.726us

**** v1   vsl|       1001 VCL_Log         c bench /16 vct 0.216us
**** v1   vsl|       1001 VCL_Log         c bench /16 vcthdrval 0.221us
**** v1   vsl|       1001 VCL_Log         c bench /16 direct 0.198us
**** v1   vsl|       1002 VCL_Log         c bench /256 vct 0.945us
**** v1   vsl|       1002 VCL_Log         c bench /256 vcthdrval 1.083us
**** v1   vsl|       1002 VCL_Log         c bench /256 direct 0.843us
**** v1   vsl|       1003 VCL_Log         c bench /1024 vct 3.247us
**** v1   vsl|       1003 VCL_Log         c bench /1024 vcthdrval 6.242us
**** v1   vsl|       1003 VCL_Log         c bench /1024 direct 2.609us
```

File: bin/varnishd/cache/cache_vrt.c (hunk 1 above)
Comment (nigoroll) [reply bin/varnishd/cache/cache_vrt.c:600]: I have pushed a variant with a separate bit in the vct table. I am not sure which other option would make sense?

**Comment by nigoroll** [comment]
bugwash consensus: would prefer an inline function, other than that we agree on the direct comparison

**Comment by nigoroll** [comment]
this ticket has served its purpose

**Pull request closed without merging**